File: c/p555.py
import sys

def mac(m, k, s, n):
    if n > m:
        return n - s
    else:
        return mac(m, k, s, mac(m, k, s, n + k))


def highest_fp(m, k, s):
    return m - 2 * s + k


def fmac(m, k, s):
    # FIXED POINT PATTERN:
    # e.g. (100, 68, 64) , [37, 38, 39, 40]
    # def f(m, k, s):
    #     return m - 2*s + k
    # This function returns the largest fixed point
    # The rest are the consecutive numbers descending from this
    # The number of fixed points equals k - s
    num_points = k - s
    if s % num_points != 0:
        return []

    highest = highest_fp(m, k, s)
    lowest = highest - (k - s) + 1
    fixed_points =  range(lowest, highest + 1)
    # print (m, k, s), ',', fixed_points
    return fixed_points
    # for i in range(m + k):
    #     if i == mac(m, k, s, i):
    #         fixed_points.append(i)

    # if fixed_points:
    #     print (m,k,s), ',', fixed_points
    # return fixed_points


def sf(m, k, s):
   return sum(fmac(m, k, s))


def s(p, m):
    return sum([sf(m, k, s) for s in range(1, p + 1) for k in range(s + 1, p + 1)])


if __name__ == '__main__':
    try:
        p, m = sys.argv[1:3]
    except:
        print "Args required: p, m"
        sys.exit(1)

    print s(int(p), int(m))

File: c/p555.c
#include <math.h>
#include <stdlib.h>
#include <stdio.h>

long mac(int m, int k, int s, int n) {
  // More like fail recursion :(
  if(n > m) {
    return n - s;
  } else {
    return mac(m, k, s, mac(m, k, s, n + k));
  }
}

long recMac(int m, int k, int s, int n) {
  // Dis is better
  int c;
  for(c = 1; c != 0; ) {
    if (n > m) {
      n -= s;
      --c;
    } else {
      n += k;
      ++c;
    }
  }
  return n;
}


long highest_fp(m, k, s) {
  return m - (2 * s) + k;
}


long sfmac(int m, int k, int s) {
  // Return sum of an array containing the fixed points
  //
  int num_points = k - s;
  if(s % num_points != 0) return 0;

  long highest = highest_fp(m, k, s);
  long lowest = highest - num_points + 1;
  long sum = 0;
  for(int i = lowest; i <= highest; i++) sum += i;
  return sum;
  
  // int fpoints[k-s];
  // int idx = 0;
  // for(int i=1; i < m + k; i++) {
  //   if(i == recMac(m, k, s, i)) {
  //     fpoints[idx] = i;
  //     ++idx;
  //   }
  // }

  // long sum = 0;
  // for(int i=0; i<idx; i++) sum += fpoints[i];
  // return sum;
}

long s(int p, int m) {
  long sum = 0;
  for(int s=1; s < p; s++) {

    for(int k=s+1; k < p + 1; k++) {
      sum += sfmac(m, k, s);
    }
  }
  return sum;
}


int main(int argc, char *argv[]) {
  int p = atoi(argv[1]);
  int m = atoi(argv[2]);
  printf("Args: %d %d\n", p, m);
  // printf("%ld\n", mac(100, 11, 10, 91));
  // printf("%ld\n", sfmac(100, 11, 10));
  printf("%ld\n", s(p, m));

  return 0;
}
